libusb/hid.c

```diff
@@ -608,102 +700,23 @@ struct hid_device_info  HID_API_EXPORT *hid_enumerate(unsigned short vendor_id,
-{
-						/*
-						This section is removed because it is too
-						invasive on the system. Getting a Usage Page
-						and Usage requires parsing the HID Report
-						descriptor. Getting a HID Report descriptor
-						involves claiming the interface. Claiming the
-						interface involves detaching the kernel driver.
-						Detaching the kernel driver is hard on the system
-						because it will unclaim interfaces (if another
-						app has them claimed) and the re-attachment of
-						the driver will sometimes change /dev entry names.
-						It is for these reasons that this section is
-						#if 0. For composite devices, use the interface
-						field in the hid_device_info struct to distinguish
-						between interfaces. */
-							unsigned char data[256];
-#ifdef DETACH_KERNEL_DRIVER
-							int detached = 0;
-							/* Usage Page and Usage */
-							res = libusb_kernel_driver_active(handle, interface_num);
-							if (res == 1) {
-								res = libusb_detach_kernel_driver(handle, interface_num);
-								if (res < 0)
-									LOG("Couldn't detach kernel driver, even though a kernel driver was attached.\n");
-								else
-									detached = 1;
-							}
-#endif
-							res = libusb_claim_interface(handle, interface_num);
-							if (res >= 0) {
-								/* Get the HID Report Descriptor. */
-								res = libusb_control_transfer(handle, LIBUSB_ENDPOINT_IN|LIBUSB_RECIPIENT_INTERFACE, LIBUSB_REQUEST_GET_DESCRIPTOR, (LIBUSB_DT_REPORT << 8)|interface_num, 0, data, sizeof(data), 5000);
-								if (res >= 0) {
-									unsigned short page=0, usage=0;
-									/* Parse the usage and usage page
-									   out of the report descriptor. */
-									get_usage(data, res,  &page, &usage);
-									cur_dev->usage_page = page;
-									cur_dev->usage = usage;
-								}
-								else
-									LOG("libusb_control_transfer() for getting the HID report failed with %d\n", res);
-
-								/* Release the interface */
-								res = libusb_release_interface(handle, interface_num);
-								if (res < 0)
-									LOG("Can't release the interface.\n");
-							}
-							else
-								LOG("Can't claim interface %d\n", res);
-#ifdef DETACH_KERNEL_DRIVER
-							/* Re-attach kernel driver if necessary. */
-							if (detached) {
-								res = libusb_attach_kernel_driver(handle, interface_num);
-								if (res < 0)
-									LOG("Couldn't re-attach kernel driver.\n");
+						#ifdef __ANDROID__
+							if (handle) {
+								/* There is (a potential) libusb Android backend, in which
+								device descriptor is not accurate up until the device is opened.
+								https://github.com/libusb/libusb/pull/874#discussion_r632801373
+								A workaround is to re-read the descriptor again.
+								Even if it is not going to be accepted into libusb master,
+								having it here won't do any harm, since reading the device descriptor
+								is as cheap as copy 18 bytes of data. */
+								libusb_get_device_descriptor(dev, &desc);
 							}
```

Comment:
this block has extra indentation - should remain at the same level as `res = libusb_open(dev, &handle);`